### PR TITLE
Watershader improvements

### PR DESCRIPTION
**Pull request by Astrak**
This PR first changes the settings of the shader_ocean example :
- reduced scene scale (there was a lot of zeros, making it difficult to understand what is at which scale, and difficult to edit)
- added dat gui (to play with the watershader parameters)

It also improves the watershader :
- added the _size_ parameter
- supports shadows
- supports toneMapping

Finally, it also changes the API : 

- instead of creating the water object, then the mirror mesh, appending the former to the latter, and using the object's material... this is hidden in `THREE.WaterMesh`, and we can now just call `water = new THREE.WaterMesh()`.

There still is the weird reflection bias to fix, not sure how to do it. But the icosahedron is now doublesided so it's not visible anymore.

### CODE REVIEW

**Comment by Mugen87** [comment]
For testing: http://rawgit.com/Astrak/three.js/watershader-improvements/examples/webgl_shaders_ocean.html

File: examples/webgl_shaders_ocean.html [hunk 1]
```diff
@@ -80,29 +81,26 @@
...
-				camera.position.set( 2000, 750, 2000 );
+				camera = new THREE.PerspectiveCamera( 55, window.innerWidth / window.innerHeight, 1, 2000 );
+				camera.position.set( 30, 30, 100 );
 
 				//
 
 				controls = new THREE.OrbitControls( camera, renderer.domElement );
-				controls.enablePan = false;
-				controls.minDistance = 1000.0;
-				controls.maxDistance = 5000.0;
```
Comment (Mugen87) [inline examples/webgl_shaders_ocean.html]: I would not delete these lines. If the user zooms too far out, the scene looks weird. Same for zooming in...

**Review by Mugen87 — CHANGES_REQUESTED** [review]

File: examples/webgl_shaders_ocean.html (hunk 1 above)
Comment (Astrak) [reply examples/webgl_shaders_ocean.html]: i re-inserted them in the latest commits, with new values since the scene scale is different.

**Comment by Mugen87** [comment]
Your implementation is okay, i just have some concerns about the missing backwards compatibility of your change. When users update to a new version of `three.js`, their existing code will definitely break. So this PR needs to be noted in the migration guide if it will be merged.

File: examples/js/WaterShader.js [hunk 2]
```diff
@@ -140,16 +140,20 @@ THREE.Water = function ( renderer, camera, scene, options ) {
...
+			'	vec4 noise = getNoise( worldPosition.xz * size );',
 			'	vec3 surfaceNormal = normalize( noise.xzy * vec3( 1.5, 1.0, 1.5 ) );',
 
 			'	vec3 diffuseLight = vec3(0.0);',
 			'	vec3 specularLight = vec3(0.0);',
 
-			'	vec3 worldToEye = eye-worldPosition;',
+			'	vec3 worldToEye = eye-worldPosition.xyz;',
 			'	vec3 eyeDirection = normalize( worldToEye );',
 			'	sunLight( surfaceNormal, eyeDirection, 100.0, 2.0, 0.5, diffuseLight, specularLight );',
```
Comment (Mugen87) [inline examples/js/WaterShader.js]: Does it still make sense to support `sunDirection` and `sunColor` when the material reacts on light source now?

Comment (Astrak) [reply examples/js/WaterShader.js]: I think it does, the bsdfs are still not used except the one of the shader, that gives his look. The new chunks are only here to support the shadowmap. However your question made me check with different lights, and the shader is currently broken if we use other types of lights than the directional lights. Working on it.

**Comment by Astrak** [comment]
How to edit the future migration guide from r87 to r88 ?

**Comment by Astrak** [comment]
The light issue is fixed now, but you can check that the lighting is still the one of the shader, it doesn't use the other provided lights, so we still need these uniforms. The main goal of this PR is just to add a native support of shadows into `THREE.Water` so we can call `water.receiveShadow = true` as any other mesh. I did not intent to change the lighting implementation here.

How does one update the migration guide ?

**Comment by Astrak** [comment]
This commit should ensure backwards compatibility : `THREE.Water` is unchanged. The new water, added on top of it with the nested objects pattern, is called `THREE.WaterMesh`.

**Comment by Mugen87** [comment]
Okay. Thanks for clarification!

**Comment by Astrak** [comment]
Maybe i could ping the original author mentionned in WaterShader.js @jbouny, to check this PR and have an advice on that reflection bias : I'm sure I had it solved in a sketch I made some time ago by reducing the scene near plane, but it doesn't change anything here, though it only is a problem at small scale... very weird to me.

**Comment by Astrak** [comment]
i think i just really made awefull things with git...

**Comment by mrdoob** [comment]
While you're at it... How about making this simpler to use by adopting `onBeforeRender`?

https://github.com/mrdoob/three.js/blob/dev/examples/js/Mirror.js#L116-L240

That way we don't need to pass `renderer`, `camera` nor `scene` in the constructor.

File: examples/js/WaterShader.js [hunk 3]
```diff
@@ -162,10 +151,11 @@ THREE.Water = function ( renderer, camera, scene, options ) {
 			'	float rf0 = 0.3;',
 			'	float reflectance = rf0 + ( 1.0 - rf0 ) * pow( ( 1.0 - theta ), 5.0 );',
 			'	vec3 scatter = max( 0.0, dot( surfaceNormal, eyeDirection ) ) * waterColor;',
-			'	vec3 albedo = mix( sunColor * diffuseLight * 0.3 + scatter, ( vec3( 0.1 ) + reflectionSample * 0.9 + reflectionSample * specularLight ), reflectance );',
+			'	vec3 albedo = mix( ( sunColor * diffuseLight * 0.3 + scatter ) * getShadowMask(), ( vec3( 0.1 ) + reflectionSample * 0.9 + reflectionSample * specularLight ), reflectance);',
```
Comment (mrdoob) [inline examples/js/WaterShader.js]: Any way the shadow could get affected by the displacement too? Right now feels like a mirror.

Comment (mrdoob) [reply examples/js/WaterShader.js]: I'll merge in the meantime though 😊

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**